src/OpenTelemetry.Api/Trace/TracerProvider.cs

```diff
@@ -43,6 +48,83 @@ protected TracerProvider()
     /// <param name="name">Name identifying the instrumentation library.</param>
     /// <param name="version">Version of the instrumentation library.</param>
     /// <returns>Tracer instance.</returns>
-    public Tracer GetTracer(string name, string? version = null)
-        => new(new ActivitySource(name ?? string.Empty, version));
+    public Tracer GetTracer(
+#if NET6_0_OR_GREATER
+        [AllowNull]
+#endif
+        string name,
+        string? version = null)
+    {
+        var tracers = this.tracers;
+        if (tracers == null)
+        {
+            // Note: Returns a no-op Tracer once dispose has been called.
+            return new(activitySource: null);
+        }
+
+        var key = new TracerKey(name, version);
+
+Retry:
+        if (!tracers.TryGetValue(key, out var tracer))
+        {
+            lock (tracers)
+            {
+                if (this.tracers == null)
+                {
+                    // Note: We check here for a race with Dispose and return a
```

Comment:
I believe we need to set `this.tracers = null` inside the same lock. Else we could still run into a situation where some thread calling `Dispose` sets `this.tracers` to `null` after this `if` check and before the new entry is added to the dictionary. We would want to return a no-op tracer in that case, but we would end up returning a valid tracer.

Reply:
I just checked it a couple times. I think it is good! Could be I'm not seeing something though. Can you write out a flow for me that you think is flawed? Here are a couple flows I'm imagining.

Case where Dispose runs in the middle of the writer and gets the lock...

* Writer thread reads the `this.tracers` on Line 58. It is valid so it begins its work.
* Dispose thread sets `this.tracers` to `null`.
* Dispose thread takes the lock.
* Reader thread misses the cache and tries to take the lock. It has to wait.
* Dispose thread finishes its clean up and releases the lock.
* Writer thread gets the lock. Now it checks `this.tracers == null`. This will be `true` now and it will return a no-op instance.

Case where Dispose runs in the middle of the writer and waits on the lock...

* Writer thread reads the `this.tracers` on Line 58. It is valid so it begins its work.
* Reader thread misses the cache and takes the lock.
* Dispose thread sets `this.tracers` to `null`.
* Dispose thread tries to takes the lock. It has to wait.
* Writer thread adds a new tracer to the cache and releases the lock. It doesn't care that `this.tracers` is now actually `null` because it is working on a local copy.
* Dispose thread gets the lock and makes all the tracers in the cache no-ops including the one that was just added.

Reply:
For case 2,
> Writer thread adds a new tracer to the cache and releases the lock. It doesn't care that this.tracers is now actually null because it is working on a local copy.

I think this is more of design choice. Yes, it doesn't care that `this.tracers` is now actually null but it could care about it 😄.

I was thinking we could offer a stronger guarantee that we would never return a `Tracer` when `TracerProvider` is disposed or being disposed. We could avoid this limbo state where the Dispose method may or may not have marked the newly returned `Tracer` no-op when its being used.

Reply:
I merged the PR because I think what's there will work well enough. I'll circle back to this comment when I have a sec to see if I can simplify it or clean it up in a way that doesn't introduce a bunch of contention.